code/modules/modular_computers/NTNet/NTNet.dm

```diff
@@ -0,0 +1,145 @@
+	// High values make displaying logs much laggier.
+	var/setting_maxlogcount = 100
+
+	// These only affect wireless. LAN (consoles) are unaffected since it would be possible to create scenario where someone turns off NTNet, and is unable to turn it back on since it refuses connections
+	var/setting_softwaredownload = 1
+	var/setting_peertopeer = 1
+	var/setting_communication = 1
+	var/setting_systemcontrol = 1
+	var/setting_disabled = 0					// Setting to 1 will disable all wireless, independently on relays status.
+
+	var/intrusion_detection_enabled = 1 		// Whether the IDS warning system is enabled
+	var/intrusion_detection_alarm = 0			// Set when there is an IDS warning due to malicious (antag) software.
+
+
+// If new NTNet datum is spawned, it replaces the old one.
+/datum/ntnet/New()
+	if(ntnet_global && (ntnet_global != src))
+		ntnet_global = src // There can be only one.
+	for(var/obj/machinery/ntnet_relay/R in machines)
+		relays.Add(R)
+		R.NTNet = src
+	build_software_lists()
+	add_log("NTNet logging system activated.")
+
+// Simplified logging: Adds a log. log_string is mandatory parameter, source is optional.
+/datum/ntnet/proc/add_log(log_string, obj/item/weapon/computer_hardware/network_card/source = null)
+	var/log_text = "[worldtime2text()] - "
+	if(source)
+		log_text += "[source.get_network_tag()] - "
+	else
+		log_text += "*SYSTEM* - "
+	log_text += log_string
+	logs.Add(log_text)
+
+
+	// We have too many logs, remove the oldest entries until we get into the limit
+	if(logs.len > setting_maxlogcount)
+		logs = logs.Copy(logs.len-setting_maxlogcount,0)
+
+
+// Checks whether NTNet operates. If parameter is passed checks whether specific function is enabled.
+/datum/ntnet/proc/check_function(specific_action = 0)
+	if(!relays || !relays.len) // No relays found. NTNet is down
+		return 0
+
+	var/operating = 0
+
+	// Check all relays. If we have at least one working relay, network is up.
+	for(var/M in relays)
+		var/obj/machinery/ntnet_relay/R = M
+		if(R.is_operational())
+			operating = 1
+			break
+
+	if(setting_disabled)
+		return 0
+
+	switch(specific_action)
+		if(NTNET_SOFTWAREDOWNLOAD)
+			return (operating && setting_softwaredownload)
+		if(NTNET_PEERTOPEER)
+			return (operating && setting_peertopeer)
+		if(NTNET_COMMUNICATION)
+			return (operating && setting_communication)
+		if(NTNET_SYSTEMCONTROL)
+			return (operating && setting_systemcontrol)
+	return operating
+
+// Builds lists that contain downloadable software.
+/datum/ntnet/proc/build_software_lists()
+	available_station_software = list()
+	available_antag_software = list()
+	for(var/F in typesof(/datum/computer_file/program))
+		var/datum/computer_file/program/prog = new F
+		// Invalid type (shouldn't be possible but just in case), invalid filetype (not executable program) or invalid filename (unset program)
+		if(!prog || !istype(prog) || prog.filename == "UnknownProgram" || prog.filetype != "PRG")
```

Comment:
@MrPerson the ```istype()``` is still useless, as only ```/datum/computer_file/program``` can be a subtype of ```/datum/computer_file/program```, by definition. even if it was ```/obj/item/weapon/potato``` with a ```parent_type``` of ```/datum/computer_file/program```, it's -still- a program (although I don't believe ```typesof()``` will return ```parent_type``` shenanigans)